File: 01_Dynamic_Programming/9251_LCS_Length_DP/Answer9251.ts
const [s1, s2] = `${require('fs').readFileSync(0)}`.split(`
`)
const s1Length = s1.length
const s2Length = s2.length
let tmpRow = new Array(s1Length + 1).fill(0)
for(let y = 1; y <= s2Length; y++){
  let diagonalX = 0
  for(let x = 1; x <= s1Length; x++){
    const upperX = tmpRow[x]
    tmpRow[x] = s1[x - 1] === s2[y - 1] ?
        diagonalX + 1 :
        tmpRow[x - 1] > tmpRow[x] ? tmpRow[x - 1] : tmpRow[x]
    diagonalX = upperX
  }
}

console.log(tmpRow[s1Length])

File: 01_Dynamic_Programming/9251_LCS_Length_DP/Answer9251.rs
fn main(){
  let l:Vec<String>=(0 ..2).map(|_|{
    let mut s=String::new();
    std::io::stdin().read_line(&mut s).unwrap();
  s}).collect();
  let (s1, s2) = (&l[0].trim(), &l[1].trim());
  let (s1_length, s2_length) = (s1.len(), s2.len());
  let mut tmp_row = vec![0; s1_length + 1];
  for y in 1 ..= s2_length {
    let mut diagonal_x = 0;
    for x in 1 ..= s1_length {
      let upper_x = tmp_row[x];
      tmp_row[x] = if s1.as_bytes()[x - 1] == s2.as_bytes()[y - 1]{
        diagonal_x + 1
      } else if tmp_row[x - 1] > tmp_row[x] {tmp_row[x - 1]} else {tmp_row[x]};
      diagonal_x = upper_x;
    }
  }
  print!("{}", tmp_row[s1_length])
}
